torchtune/modules/transformer.py

```diff
@@ -460,8 +465,14 @@ def forward(
         # shape: [b, s, d]
         h = self.norm(h)
 
-        # shape: [b, s, out_dim] - out_dim is usually the vocab size
-        output = self.output(h).float()
+        # shape: [b, seq_len//num_chunks, out_dim] - out_dim is usually the vocab size
+        if self.num_output_chunks > 0:
+            # chunk for ChunkedCrossEntropyLoss, saving memory
```

Comment:
My hypothesis was that by chunking before the linear, the output chunks are not views into the same underlying tensor allowing earlier ones to be freed and have their memory reused for later ones.

I wonder if we can check the memory snapshot to confirm that if not already, as it might be nice to understand the constraint on the modeling side (namely, having to chunk the _input_ to the final output projection rather than chunking the output itself).

Reply:
This has been bugging me for a while and I still don't understand this. Where would they be freed though? We pass the entire list of chunks to the loss? Or did I misunderstand your point?

Reply:
its not the the chunks are fred. If we split AFTER, then we have chunks + whole output. When we split BEFORE, we have only the chunks.

The memory savings is because, when we have chunks, the extra memory by upcasting is at most of size of one chunk, instead of doubling the whole output.

Ideally, if we split AFTER, pytorch should learn to delete the whole output that we have no need for anymore. But this doesnt happen.

![image](https://github.com/user-attachments/assets/5ba2670b-8326-4eb6-8190-e843d3ede0c8)
![image](https://github.com/user-attachments/assets/b8a9670b-601d-4d21-a13f-e4351dbf9f5e)